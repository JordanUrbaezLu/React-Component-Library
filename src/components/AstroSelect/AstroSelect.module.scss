.astroSelect {
  border-radius: 5px;
  outline: none !important;
  padding: 5px;
  font-size: 28px;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
  display: flex;
  flex-direction: row;
  align-items: center;
  cursor: pointer;
  position: relative;
}

.selectMenu {
    display: flex;
    flex-direction: column;
    position: absolute;
}

.primary {
  border: 3px solid rgb(255, 220, 100);
  &:focus {
    border: 3px solid rgb(255, 183, 77);
  }
}

.secondary {
  border: 3px solid rgb(192, 137, 243);
  &:focus {
    border: 3px solid rgb(147 51 234);
  }
}

.iconPrimary {
    color: rgb(255, 183, 77);
    display: flex;
}

.iconSecondary {
    color: rgb(147 51 234);
    display: flex;
}