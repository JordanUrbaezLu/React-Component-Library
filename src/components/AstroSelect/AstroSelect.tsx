import classNames from "classnames";
import styles from "./AstroSelect.module.scss";
import itemStyles from "./AstroSelectItem/AstroSelectItem.module.scss";
import { IoChevronDownSharp, IoChevronUpSharp } from "react-icons/io5";
import * as React from "react";

export type AstroSelectVariant = "primary" | "secondary";
export type AstroSelectSize = "small" | "medium" | "large";

interface AstroSelectProps {
  variant?: AstroSelectVariant;
  children?: React.ReactNode;
  size?: AstroSelectSize;
  width?: number;
  label?: string;
}

const AstroSelect: React.FC<AstroSelectProps> = ({
  children,
  label,
  width,
  variant = "primary",
  size = "medium",
}) => {
  const containerClasses = classNames(
    styles.astroSelectContainer,
    size === "small" && styles.smallContainer,
    size === "medium" && styles.mediumContainer,
    size === "large" && styles.largeContainer
  );

  const labelClasses = classNames(
    styles.label,
    size === "small" && styles.smallLabel,
    size === "medium" && styles.mediumLabel,
    size === "large" && styles.largeLabel,
    variant === "primary" && styles.primaryLabel,
    variant === "secondary" && styles.secondaryLabel
  );

  const classes = classNames(
    styles.astroSelect,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    size === "small" && styles.small,
    size === "medium" && styles.medium,
    size === "large" && styles.large
  );

  const iconClasses = classNames(
    styles.icon,
    variant === "primary" && styles.primaryIcon,
    variant === "secondary" && styles.secondaryIcon,
    size === "small" && styles.smallIcon,
    size === "medium" && styles.mediumIcon,
    size === "large" && styles.largeIcon
  );

  const itemClasses = classNames(
    itemStyles.astroSelectItem,
    size === "small" && itemStyles.small,
    size === "medium" && itemStyles.medium,
    size === "large" && itemStyles.large
  );

  const [isOpen, setIsOpen] = React.useState<boolean>(false);
  const [value, setValue] = React.useState<string>("");

  return (
    <div
      className={containerClasses}
      style={width ? { width: `${width}px` } : {}}
    >
      {label && <div className={labelClasses}>{label}</div>}
      <div className={classes}>
        <div style={{ paddingLeft: "4px" }}>{value}</div>
        <div className={iconClasses} onClick={() => setIsOpen(!isOpen)}>
          {isOpen ? <IoChevronUpSharp /> : <IoChevronDownSharp />}
        </div>
      </div>
      {isOpen && (
        // TODO: Have menu hover over content, currently display block and moves everything below furthur down
        <div className={styles.selectOptionsMenu}>
          {React.Children.map(children, (child) => {
            const item = child as React.ReactElement;
            return (
              <div
                onClick={() => {
                  setValue(item.props.value);
                  setIsOpen(!isOpen);
                }}
                className={itemClasses}
              >
                {item}
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
};

export default AstroSelect;
