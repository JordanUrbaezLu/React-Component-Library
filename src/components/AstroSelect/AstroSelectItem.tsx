import * as React from "react";

interface AstroSelectItemProps {
  children?: React.ReactNode;
}

const AstroSelectItem: React.FC<AstroSelectItemProps> = ({ children }) => {
  return <div>{children}</div>;
};

export default AstroSelectItem;
