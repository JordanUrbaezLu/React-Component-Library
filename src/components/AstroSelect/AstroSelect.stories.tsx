import { ComponentStory, ComponentMeta } from "@storybook/react";
import AstroSelect from "./AstroSelect";
import AstroSelectItem from "./AstroSelectItem/AstroSelectItem";

export default {
  title: "AstroSelect",
  component: AstroSelect,
} as ComponentMeta<typeof AstroSelect>;

const Template: ComponentStory<typeof AstroSelect> = (args) => {
  return (
    <AstroSelect {...args}>
      <AstroSelectItem value="30" />
      <AstroSelectItem value="40" />
      <AstroSelectItem value="50" />
    </AstroSelect>
  );
};

export const Default = Template.bind({});


