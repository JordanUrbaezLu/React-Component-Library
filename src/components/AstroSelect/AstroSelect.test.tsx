import { render, screen } from "@testing-library/react";
import AstroSelect, {
  AstroSelectVariant,
  AstroSelectSize,
} from "./AstroSelect";
import { axe, toHaveNoViolations } from "jest-axe";
import "@testing-library/jest-dom";
expect.extend(toHaveNoViolations);

describe("AstroSelect", () => {
  test("Should render component correctly.", () => {
    const { container } = render(<AstroSelect />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(<AstroSelect />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

// TODO: Finish test for AstroSelect once label prop is added in AstroSelect.tsx
