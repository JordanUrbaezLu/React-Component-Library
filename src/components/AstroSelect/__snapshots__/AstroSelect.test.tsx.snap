// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstroSelect Should render component correctly. 1`] = `
<div
  class="astroSelectContainer mediumContainer"
>
  <div
    class="astroSelect primary medium"
  >
    <div
      style="padding-left: 4px;"
    />
    <div
      class="icon primaryIcon mediumIcon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M112 184l144 144 144-144"
          fill="none"
          stroke-linecap="square"
          stroke-miterlimit="10"
          stroke-width="48"
        />
      </svg>
    </div>
  </div>
</div>
`;
